cpp/src/arrow/compute/exec/partition_util.cc

```diff
@@ -0,0 +1,91 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/exec/partition_util.h"
+#include <mutex>
+
+namespace arrow {
+namespace compute {
+
+PartitionLocks::PartitionLocks()
+    : num_prtns_(0),
+      locks_(nullptr),
+      rand_seed_{0, 0, 0, 0, 0, 0, 0, 0},
+      rand_engine_(rand_seed_) {}
+
+PartitionLocks::~PartitionLocks() { CleanUp(); }
+
+void PartitionLocks::Init(int num_prtns) {
+  num_prtns_ = num_prtns;
+  locks_ = new PartitionLock[num_prtns];
+  for (int i = 0; i < num_prtns; ++i) {
+    locks_[i].lock.store(false);
+  }
+}
+
+void PartitionLocks::CleanUp() {
+  if (locks_) {
+    delete[] locks_;
+    locks_ = nullptr;
+  }
+  num_prtns_ = 0;
+}
+
+std::atomic<bool>* PartitionLocks::lock_ptr(int prtn_id) {
```

Comment:
I actually like pointer here more because you never want to make a copy, and I don't think there's ever a way to enforce someone in the future to _only_ take a reference, and never copy by value.

Reply:
In case of atomic I also like pointers better. The intrinsics for atomic instructions take a pointer to the memory address. And atomics are not copy-able.

Reply:
Any attempted copy would be a compile error.  If you think the style is better with a pointer that is fine.